documentation-site/components/yard/code-generator.ts

```diff
@@ -1,77 +1,251 @@
+  const value = prop.value;
+  switch (prop.type as PropTypes) {
+    case PropTypes.String:
+      return t.stringLiteral(value);
+    case PropTypes.Boolean:
+      return t.jsxExpressionContainer(t.booleanLiteral(value));
+    case PropTypes.Ref:
+      return null;
+    case PropTypes.Number:
+    case PropTypes.Array:
+    case PropTypes.Object:
+    case PropTypes.Function:
+    case PropTypes.ReactNode:
+      return t.jsxExpressionContainer((template.ast(value) as any).expression);
+    case PropTypes.Enum:
+      return t.jsxExpressionContainer(t.identifier(value));
+    case PropTypes.Overrides:
+      const activeValues = Object.entries(value).filter(
+        ([, val]: any) => val.active,
+      );
+      if (activeValues.length === 0) return null;
+      const keys = activeValues.map(([key, val]: [string, any]) =>
+        t.objectProperty(
+          t.identifier(key),
+          t.objectExpression([
+            t.objectProperty(t.identifier('style'), template.expression(
+              val.style,
+            )({}) as any),
+          ]),
+        ),
+      );
+      return t.jsxExpressionContainer(t.objectExpression(keys));
+  }
+};
+
+const getAstReactHooks = (props: {[key: string]: TProp}) => {
+  const hooks: babel.types.ExpressionStatement[] = [];
+  const buildReactHook = template(
+    `const [%%name%%, %%setName%%] = React.useState(%%value%%);`,
+  );
+  Object.keys(props).forEach(name => {
+    if (props[name].meta && (props[name] as any).meta.stateful === true) {
+      hooks.push(buildReactHook({
+        name: t.identifier(name),
+        setName: t.identifier(`set${name[0].toUpperCase() + name.slice(1)}`),
+        value: getAstPropValue(props[name]),
+      }) as any);
+    }
+  });
+  return hooks;
+};
+
+const getAstImport = (identifiers: string[], source: string) => {
+  return t.importDeclaration(
+    identifiers.map(identifier =>
+      t.importSpecifier(t.identifier(identifier), t.identifier(identifier)),
+    ),
+    t.stringLiteral(source),
+  );
+};
+
+const getEnumsToImport = (props: {[key: string]: TProp}) => {
+  const enums: string[] = [];
+  Object.keys(props).forEach(name => {
+    if (props[name].type === PropTypes.Enum && props[name].value) {
+      enums.push(name.toUpperCase());
+    }
+  });
+  return enums;
+};
+
+const getAstJsxElement = (
+  name: string,
+  attrs: (t.JSXAttribute | null)[],
+  children: TJsxChild[],
+) => {
+  const isSelfClosing = children.length === 0;
+  return t.jsxElement(
+    t.jsxOpeningElement(
+      t.jsxIdentifier(name),
+      attrs.filter(attr => !!attr) as t.JSXAttribute[],
+      isSelfClosing,
+    ),
+    isSelfClosing ? null : t.jsxClosingElement(t.jsxIdentifier(name)),
+    children,
+    true,
+  );
+};
+
+const getAstThemeImport = (isCustomTheme: boolean, themePrimitives: string) => {
+  if (!isCustomTheme) return [];
+  const buildImportTheme = template(
+    `import {ThemeProvider, createTheme, %%primitives%%} from "baseui"`,
```

Comment:
It would not work for `template()` since that returns a function (so the value would end up in its closure?) but there is a similar `template.ast()` API that directly returns ast. 

I guess the argument for `template()` is performance - you don't need to parse the string every single time, just once and then it only injects the variable into the created AST.

Reply:
> Also, does template eventually consider this an import statement node

Yea, it gets the correct type. It's identical to AST that you would create manually with babel-types.